hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java

```diff
@@ -143,12 +147,124 @@ public void markBlocksAsDeleted(KeyValueContainerData data, Collection<Long> del
+  }
+
+  private ContainerDiff compareContainerMerkleTree(ContainerProtos.ContainerMerkleTree thisMerkleTree,
+                                                   ContainerProtos.ContainerMerkleTree peerMerkleTree) {
+
+    ContainerDiff report = new ContainerDiff();
+    if (thisMerkleTree.getDataChecksum() == peerMerkleTree.getDataChecksum()) {
+      return new ContainerDiff();
+    }
+
+    List<ContainerProtos.BlockMerkleTree> thisBlockMerkleTreeList = thisMerkleTree.getBlockMerkleTreeList();
+    List<ContainerProtos.BlockMerkleTree> peerBlockMerkleTreeList = peerMerkleTree.getBlockMerkleTreeList();
+    // Since we are reconciling our container with a peer, We only need to go through the peer's block list
+    int peerIdx = 0, thisIdx = 0;
+    while (peerIdx < peerBlockMerkleTreeList.size() || thisIdx < thisBlockMerkleTreeList.size()) {
+      ContainerProtos.BlockMerkleTree thisBlockMerkleTree = thisIdx < thisBlockMerkleTreeList.size() ?
+          thisBlockMerkleTreeList.get(thisIdx) : null;
+      ContainerProtos.BlockMerkleTree peerBlockMerkleTree = peerIdx < peerBlockMerkleTreeList.size() ?
+          peerBlockMerkleTreeList.get(peerIdx) : null;
+
+      // We have checked all the peer blocks, So we can return the ContainerDiff report
+      if (peerBlockMerkleTree == null) {
+        return report;
+      }
+
+      // peerBlockMerkleTree is not null, but thisBlockMerkleTree is null. Means we have missing blocks.
+      if (thisBlockMerkleTree == null) {
+        report.addMissingBlock(peerBlockMerkleTree);
+        peerIdx++;
+        continue;
+      }
+
+      // BlockId matches for both thisBlockMerkleTree and peerBlockMerkleTree
+      if (thisBlockMerkleTree.getBlockID() == peerBlockMerkleTree.getBlockID()) {
+        if (thisBlockMerkleTree.getBlockChecksum() != peerBlockMerkleTree.getBlockChecksum()) {
+          compareBlockMerkleTree(report, thisBlockMerkleTree, peerBlockMerkleTree);
+        }
+        peerIdx++;
+        thisIdx++;
+      } else {
+        if (peerBlockMerkleTree.getBlockID() > thisBlockMerkleTree.getBlockID()) {
+          thisIdx++;
+        } else {
+          report.addMissingBlock(peerBlockMerkleTree);
+          peerIdx++;
+        }
+      }
+    }
+    return report;
+  }
+
+  private void compareBlockMerkleTree(ContainerDiff report, ContainerProtos.BlockMerkleTree thisBlockMerkleTree,
+                                      ContainerProtos.BlockMerkleTree peerBlockMerkleTree) {
+
+    List<ContainerProtos.ChunkMerkleTree> thisChunkMerkleTreeList = thisBlockMerkleTree.getChunkMerkleTreeList();
+    List<ContainerProtos.ChunkMerkleTree> peerChunkMerkleTreeList = peerBlockMerkleTree.getChunkMerkleTreeList();
+
+    // Since we are reconciling our container with a peer, We only need to go through the peer's chunk list
+    int peerIdx = 0, thisIdx = 0;
+    while (peerIdx < peerChunkMerkleTreeList.size() || thisIdx < thisChunkMerkleTreeList.size()) {
```

Comment:
Can simplify this to the more "traditional" way to merge two sorted lists? This will help future developers recognize what is happening here.

IME there is usually one while loop handling the case where the lists interleave. This is followed by two more loops to handle the remainder of either list. This differs from what is here where there is one loop and null is being used as a placeholder to handle the list remainders.

Reply:
Updated the implementation